# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/media/MediaDecoderStateMachine.cpp
# Commit: 6a50de183c32
# Full Hash: 6a50de183c329b420dc894542d413ea343503383
# Author: alwu <alwu@mozilla.com>
# Date: 2022-11-26 21:27:35
# Description:
#   Bug 1801692 - always check other conditions before requesting data in order to ensure that we don't request data during another request or seeking. r=padenot
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D162790
# ==============================================================================

diff -r 738ee02dfc2a -r 6a50de183c32 dom/media/MediaDecoderStateMachine.cpp
--- a/dom/media/MediaDecoderStateMachine.cpp	Fri Nov 25 23:08:46 2022 +0000
+++ b/dom/media/MediaDecoderStateMachine.cpp	Fri Nov 25 23:27:22 2022 +0000
@@ -716,6 +716,24 @@
     }
   }
 
+  bool ShouldRequestNextKeyFrame() const {
+    if (!mVideoFirstLateTime) {
+      return false;
+    }
+    const double elapsedTimeMs =
+        (TimeStamp::Now() - *mVideoFirstLateTime).ToMilliseconds();
+    const bool rv = elapsedTimeMs >=
+                    StaticPrefs::media_decoder_skip_when_video_too_slow_ms();
+    if (rv) {
+      PROFILER_MARKER_UNTYPED("Skipping to next keyframe", MEDIA_PLAYBACK);
+      SLOG(
+          "video has been late behind media time for %f ms, should skip to "
+          "next key frame",
+          elapsedTimeMs);
+    }
+    return rv;
+  }
+
  private:
   void DispatchDecodeTasksIfNeeded();
   void MaybeStartBuffering();
@@ -787,24 +805,6 @@
         [this]() { mDormantTimer.CompleteRequest(); });
   }
 
-  bool ShouldRequestNextKeyFrame() const {
-    if (!mVideoFirstLateTime) {
-      return false;
-    }
-    const double elapsedTimeMs =
-        (TimeStamp::Now() - *mVideoFirstLateTime).ToMilliseconds();
-    const bool rv = elapsedTimeMs >=
-                    StaticPrefs::media_decoder_skip_when_video_too_slow_ms();
-    if (rv) {
-      PROFILER_MARKER_UNTYPED("Skipping to next keyframe", MEDIA_PLAYBACK);
-      SLOG(
-          "video has been late behind media time for %f ms, should skip to "
-          "next key frame",
-          elapsedTimeMs);
-    }
-    return rv;
-  }
-
   // Time at which we started decoding.
   TimeStamp mDecodeStartTime;
 
@@ -1199,17 +1199,57 @@
 
   void HandleError(const MediaResult& aError, bool aIsAudio);
 
+  bool ShouldRequestData(MediaData::Type aType) const {
+    MOZ_DIAGNOSTIC_ASSERT(aType == MediaData::Type::AUDIO_DATA ||
+                          aType == MediaData::Type::VIDEO_DATA);
+    if (aType == MediaData::Type::AUDIO_DATA &&
+        (mAudioSeekRequest.Exists() || mAudioDataRequest.Exists() ||
+         IsDataWaitingForTimestampAdjustment(MediaData::Type::AUDIO_DATA))) {
+      return false;
+    }
+    if (aType == MediaData::Type::VIDEO_DATA &&
+        (mVideoSeekRequest.Exists() || mVideoDataRequest.Exists() ||
+         IsDataWaitingForTimestampAdjustment(MediaData::Type::VIDEO_DATA))) {
+      return false;
+    }
+    return true;
+  }
+
+  void HandleAudioCanceled() override {
+    if (ShouldRequestData(MediaData::Type::AUDIO_DATA)) {
+      mMaster->RequestAudioData();
+    }
+  }
+
+  void HandleAudioWaited(MediaData::Type aType) override {
+    if (ShouldRequestData(MediaData::Type::AUDIO_DATA)) {
+      mMaster->RequestAudioData();
+    }
+  }
+
+  void HandleVideoCanceled() override {
+    if (ShouldRequestData(MediaData::Type::VIDEO_DATA)) {
+      mMaster->RequestVideoData(mMaster->GetMediaTime(),
+                                ShouldRequestNextKeyFrame());
+    };
+  }
+
+  void HandleVideoWaited(MediaData::Type aType) override {
+    if (ShouldRequestData(MediaData::Type::VIDEO_DATA)) {
+      mMaster->RequestVideoData(mMaster->GetMediaTime(),
+                                ShouldRequestNextKeyFrame());
+    };
+  }
+
   void EnsureAudioDecodeTaskQueued() override {
-    if (mAudioSeekRequest.Exists() || mAudioDataRequest.Exists() ||
-        IsDataWaitingForTimestampAdjustment(MediaData::Type::AUDIO_DATA)) {
+    if (!ShouldRequestData(MediaData::Type::AUDIO_DATA)) {
       return;
     }
     DecodingState::EnsureAudioDecodeTaskQueued();
   }
 
   void EnsureVideoDecodeTaskQueued() override {
-    if (mVideoSeekRequest.Exists() || mVideoDataRequest.Exists() ||
-        IsDataWaitingForTimestampAdjustment(MediaData::Type::VIDEO_DATA)) {
+    if (!ShouldRequestData(MediaData::Type::VIDEO_DATA)) {
       return;
     }
     DecodingState::EnsureVideoDecodeTaskQueued();
